zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,924 @@
+            .map(|input| match input {
+                transparent::Input::PrevOut { sequence, .. } => sequence,
+                transparent::Input::Coinbase { sequence, .. } => sequence,
+            })
+            .try_for_each(|sequence| (&mut hash).write_u32::<LittleEndian>(*sequence))?;
+
+        writer.write_all(hash.finalize().as_ref())
+    }
+
+    /// Writes the u256 hash of the transactions outputs to the provided Writer
+    fn hash_outputs<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        if self.hash_type.masked() != HashType::SINGLE && self.hash_type.masked() != HashType::NONE
+        {
+            self.outputs_hash(writer)
+        } else if self.hash_type.masked() == HashType::SINGLE
+            && self
+                .input
+                .as_ref()
+                .map(|&(_, _, index)| index)
+                .map(|index| index < self.trans.outputs().len())
+                .unwrap_or(false)
+        {
+            self.single_output_hash(writer)
+        } else {
+            writer.write_all(&[0; 32])
+        }
+    }
+
+    fn outputs_hash<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(ZCASH_OUTPUTS_HASH_PERSONALIZATION)
+            .to_state();
+
+        self.trans
+            .outputs()
+            .iter()
+            .try_for_each(|output| output.zcash_serialize(&mut hash))?;
+
+        writer.write_all(hash.finalize().as_ref())
+    }
+
+    fn single_output_hash<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        let &(_, _, output) = self
+            .input
+            .as_ref()
+            .expect("already checked index is some in `hash_outputs`");
+
+        let output = &self.trans.outputs()[output];
+
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(ZCASH_OUTPUTS_HASH_PERSONALIZATION)
+            .to_state();
+
+        output.zcash_serialize(&mut hash)?;
+
+        writer.write_all(hash.finalize().as_ref())
+    }
+
+    fn hash_joinsplits<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        let has_joinsplits = match self.trans {
+            Transaction::V1 { .. } => false,
+            Transaction::V2 { joinsplit_data, .. } | Transaction::V3 { joinsplit_data, .. } => {
+                joinsplit_data.is_some()
+            }
+            Transaction::V4 { joinsplit_data, .. } => joinsplit_data.is_some(),
+        };
+
+        if !has_joinsplits {
+            return writer.write_all(&[0; 32]);
+        }
+
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(ZCASH_JOINSPLITS_HASH_PERSONALIZATION)
+            .to_state();
+
+        match self.trans {
+            Transaction::V2 {
+                joinsplit_data: Some(jsd),
+                ..
+            }
+            | Transaction::V3 {
```

Comment:
Already checked for V2

```suggestion
            Transaction::V3 {
```

Reply:
this change wont compile because it makes a new branch arm for v2 without supplying a body.

Reply:
oh wait, i see what you're saying, v2 isn't a thing, got it.

Reply:
Yeah sorry the diff and comment aren't clear.